bfsscp/bfsscp.go

```diff
@@ -0,0 +1,376 @@
+// Package bfsscp abstracts an SSH/SCP workflow
+//
+// When imported, it registers both `scp://` and `ssh://` global scheme resolver and can be used like:
+//
+//   import (
+//     "github.com/bsm/bfs"
+//
+//     _ "github.com/bsm/bfs/bfsscp"
+//   )
+//
+//   func main() {
+//     ctx := context.Background()
+//     b, _ := bfs.Connect(ctx, "ssh://user:pass@hostname:22/path/to/root?tmpdir=%2Fcustom%2Ftmp")
+//     f, _ := b.Open(ctx, "file/within/root.txt")
+//     ...
+//   }
+//
+// bfs.Connect supports the following query parameters:
+//
+//   tmpdir - custom temp dir
+//
+package bfsscp
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bmatcuk/doublestar"
+	"github.com/bsm/bfs"
+	"github.com/bsm/bfs/internal"
+	"github.com/kr/fs"
+	"github.com/pkg/sftp"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/ssh"
+)
+
+// Register allows for registering more schemes with SCP
+func Register(_ context.Context, u *url.URL) (bfs.Bucket, error) {
+	query := u.Query()
+	address := net.JoinHostPort(u.Host, u.Port())
+
+	username, password := "", ""
+	if u.User != nil {
+		username = u.User.Username()
+		password, _ = u.User.Password()
+	}
+
+	return New(address, &Config{
+		Username: username,
+		Password: password,
+		Prefix:   u.Path,
+		TempDir:  query.Get("tmpdir"),
+	})
+}
+
+func init() {
+	bfs.Register("ssh", Register)
+	bfs.Register("scp", Register)
+}
+
+// Config is passed to New to configure the SSH connection.
+type Config struct {
+	// Username to use.
+	Username string
+	// Password to use in combination with Username.
+	Password string
+	// An optional path prefix.
+	Prefix string
+	// A custom temp dir.
+	TempDir string
+}
+
+func (c *Config) norm() error {
+	c.Prefix = strings.TrimPrefix(c.Prefix, "/")
+	if c.Prefix != "" && !strings.HasSuffix(c.Prefix, "/") {
+		c.Prefix = c.Prefix + "/"
+	}
+	return nil
+}
+
+type bucket struct {
+	conn   *ssh.Client
+	client *sftp.Client
+	config *Config
+}
+
+// New initiates an bfs.Bucket backed by ssh.
+func New(address string, cfg *Config) (bfs.Bucket, error) {
```

Comment:
can we accept a context hare?

Reply:
The BFS standard seems to be passing `ctx` only on the Glob/Head/Open/Create/Remove.
If we added it here, we'd probably want to change it elsewhere too.

Reply:
k, leave it for now